storage/src/mmr/mutable.rs

```diff
@@ -140,61 +162,78 @@ impl<K: Array, V: Array, H: CHasher> MutableMmr<K, V, H> {
 
     /// Get the number of operations that have been applied to this store.
     pub fn op_count(&self) -> u64 {
-        self.log.len() as u64
+        self.log.len() as u64 + self.pruned_loc
     }
 
     /// Updates `key` to have value `value`.  If the key already has this same value, then this is a
     /// no-op.
+    ///
+    /// Also move exactly one active operation to tip if at least one is movable.
     pub fn update(&mut self, hasher: &mut H, key: K, value: V) {
-        let new_loc = self.log.len();
+        let new_loc = self.log.len() as u64 + self.pruned_loc;
 
         // Update the snapshot.
         if let Some(loc) = self.snapshot.get_mut(&key) {
-            let last_value = match self.log[*loc].op_type {
+            let i = *loc - self.pruned_loc;
+            let last_value = match self.log[i as usize].op_type {
                 Type::Deleted => panic!("deleted key should not be in snapshot: {}", key),
                 Type::Update(ref v) => v,
             };
             if value == *last_value {
                 // Trying to assign the same value is a no-op.
                 return;
             }
+            if *loc == self.inactivity_floor_loc {
```

Comment:
I recommend avoiding handling this out-of-band here.

Reply:
removed